### PR TITLE
Validate that blockperiodseconds is set to a positive integer

### PR DESCRIPTION
**Pull request by georgep9**
## PR description

Implemented getPositiveInt function in JsonUtil to validate a JSON positive number value. Using this function, the blockperiodseconds should now be validated whenever retrieved from the genensis config. 

## Fixed Issue(s)

Fixes https://github.com/hyperledger/besu/issues/3093

### CODE REVIEW

File: config/src/test/java/org/hyperledger/besu/config/JsonBftConfigOptionsTest.java [hunk 1]
```diff
@@ -70,6 +71,13 @@ public void shouldGetDefaultBlockPeriodFromDefaultConfig() {
         .isEqualTo(EXPECTED_DEFAULT_BLOCK_PERIOD);
   }
 
+  @Test
```
Comment (lucassaldanha) [inline config/src/test/java/org/hyperledger/besu/config/JsonBftConfigOptionsTest.java:74]: Do we wanna add some extra edge cases like negative numbers?

File: besu/src/main/java/org/hyperledger/besu/cli/subcommands/operator/GenerateBlockchainConfig.java [hunk 2]
```diff
@@ -136,8 +136,8 @@ private void checkPreconditions() {
   /** Generates output directory with all required configuration files. */
   private void generateBlockchainConfig() {
     try {
-      handleOutputDirectory();
       parseConfig();
+      handleOutputDirectory();
```
Comment (siladu) [inline besu/src/main/java/org/hyperledger/besu/cli/subcommands/operator/GenerateBlockchainConfig.java:140]: I would postpone adding this to the PR (unless we end up modifying besu operator in this PR as well). This change would be better isolated in it's own PR so we don't have to worry about testing alongside the other stuff.

Comment (georgep9) [reply besu/src/main/java/org/hyperledger/besu/cli/subcommands/operator/GenerateBlockchainConfig.java:140]: Definitely

File: config/src/test/java/org/hyperledger/besu/config/JsonBftConfigOptionsTest.java [hunk 3]
```diff
@@ -70,6 +71,20 @@ public void shouldGetDefaultBlockPeriodFromDefaultConfig() {
...
+  public void shouldThrowOnDecimalBlockPeriod() {
+    final BftConfigOptions config = fromConfigOptions(singletonMap("BlockPeriodSeconds", 1.5));
+    assertThatThrownBy(() -> config.getBlockPeriodSeconds())
+        .isInstanceOf(IllegalArgumentException.class);
+  }
+
+  @Test
+  public void shouldThrowOnNonPositiveBlockPeriod() {
```
Comment (siladu) [inline config/src/test/java/org/hyperledger/besu/config/JsonBftConfigOptionsTest.java:82]: I think these would better live in PositiveNumberTest. 
But then you might still have one in the *ConfigOptionsTest classes that's something like "shouldThrowForNonPositiveInteger"

File: config/src/test/java/org/hyperledger/besu/config/CliqueConfigOptionsTest.java [hunk 4]
```diff
@@ -64,6 +65,20 @@ public void shouldGetDefaultBlockPeriodFromDefaultConfig() {
...
+    assertThatThrownBy(() -> config.getBlockPeriodSeconds())
+        .isInstanceOf(IllegalArgumentException.class);
+  }
+
+  @Test
+  public void shouldThrowOnNonPositiveBlockPeriod() {
+    final CliqueConfigOptions config = fromConfigOptions(singletonMap("BlockPeriodSeconds", -1));
+    assertThatThrownBy(() -> config.getBlockPeriodSeconds())
+        .isInstanceOf(IllegalArgumentException.class);
+  }
```
Comment (siladu) [inline config/src/test/java/org/hyperledger/besu/config/CliqueConfigOptionsTest.java:80]: I think these would better live in PositiveNumberTest.
But then you might still have one in the *ConfigOptionsTest classes that's something like "shouldThrowForNonPositiveInteger"

File: util/src/test/java/org/hyperledger/besu/util/number/PositiveNumberTest.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (siladu) [inline util/src/test/java/org/hyperledger/besu/util/number/PositiveNumberTest.java:2]: Well spotted for noticing the copyright was required!
I think it changed quite recently (some docs may be out of date?) so it should now read "Copyright Hyperledger Besu Contributors."

File: config/src/test/java/org/hyperledger/besu/config/JsonBftConfigOptionsTest.java [hunk 6]
```diff
@@ -70,6 +71,13 @@ public void shouldGetDefaultBlockPeriodFromDefaultConfig() {
         .isEqualTo(EXPECTED_DEFAULT_BLOCK_PERIOD);
   }
 
+  @Test
+  public void shouldThrowOnNonPositiveBlockPeriod() {
+    final BftConfigOptions config = fromConfigOptions(singletonMap("BlockPeriodSeconds", -1));
```
Comment (siladu) [inline config/src/test/java/org/hyperledger/besu/config/JsonBftConfigOptionsTest.java:76]: nit: could this be "blockperiodseconds" for consistency? Don't think lowercase matters here but it does in other config-related things so better to keep it lower everywhere I think.

File: config/src/test/java/org/hyperledger/besu/config/CliqueConfigOptionsTest.java [hunk 7]
```diff
@@ -64,6 +65,13 @@ public void shouldGetDefaultBlockPeriodFromDefaultConfig() {
         .isEqualTo(EXPECTED_DEFAULT_BLOCK_PERIOD);
   }
 
+  @Test
+  public void shouldThrowOnNonPositiveBlockPeriod() {
+    final CliqueConfigOptions config = fromConfigOptions(singletonMap("BlockPeriodSeconds", -1));
```
Comment (siladu) [inline config/src/test/java/org/hyperledger/besu/config/CliqueConfigOptionsTest.java:70]: nit: suggest "blockperiodseconds" in lowercase

File: config/src/main/java/org/hyperledger/besu/config/JsonUtil.java [hunk 8]
```diff
@@ -291,4 +293,15 @@ private static boolean validateInt(final JsonNode node) {
     }
     return true;
   }
+
+  public static int getPositiveInt(
```
Comment (siladu) [inline config/src/main/java/org/hyperledger/besu/config/JsonUtil.java:297]: nit: Could this method live next to getInt? It's very related to it

File: util/src/test/java/org/hyperledger/besu/util/number/PositiveNumberTest.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    final String positiveNumberString = "1";
+    final int positiveNumberInt = 1;
+    PositiveNumber positiveNumber = PositiveNumber.fromString(positiveNumberString);
+    assertThat(positiveNumber).isNotNull();
+    assertThat(positiveNumber.getValue()).isEqualTo(positiveNumberInt);
+  }
+
+  @Test
+  public void shouldThrowOnDecimalValueFromString() {
+    assertThatThrownBy(() -> PositiveNumber.fromString("0.1"))
```
Comment (siladu) [inline util/src/test/java/org/hyperledger/besu/util/number/PositiveNumberTest.java:35]: "1.1" is a better choice here as "0.1" might fail due to being non-positive, rather than a decimal.

**Review by siladu — CHANGES_REQUESTED** [review]
Minor nitpicks and suggestions in the comments.

I am going to be annoying though and ask that you make the same change for `BftFork.getBlockPeriodSeconds` please :)
It is possible to use a 'transition' to change the blockperiodseconds of an existing network so it needs the same validation constraints applied to it.
See: https://besu.hyperledger.org/en/stable/HowTo/Configure/Consensus-Protocols/IBFT/#configure-block-time-on-an-existing-network-deployment

**Review by siladu — CHANGES_REQUESTED** [review]

**Review by siladu — COMMENTED** [review]
Code looks good 👍 . 
I want to have a play with it locally before I approve which will probably be next year :)

**Review by siladu — APPROVED** [review]
LGTM!
